presto-spi/src/main/java/io/prestosql/spi/block/MapHashTables.java

```diff
@@ -15,23 +15,128 @@
 
 import io.prestosql.spi.PrestoException;
 import io.prestosql.spi.type.MapType;
+import org.openjdk.jol.info.ClassLayout;
 
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.Arrays;
+import java.util.Optional;
+
+import static io.airlift.slice.SizeOf.sizeOf;
 import static io.prestosql.spi.StandardErrorCode.NOT_SUPPORTED;
+import static java.lang.String.format;
 
-final class MapHashTables
+@ThreadSafe
+public final class MapHashTables
 {
-    private MapHashTables() {}
+    public static final int INSTANCE_SIZE = ClassLayout.parseClass(MapHashTables.class).instanceSize();
+
+    // inverse of hash fill ratio, must be integer
+    static final int HASH_MULTIPLIER = 2;
+
+    private final MapType mapType;
+
+    @SuppressWarnings("VolatileArrayField")
+    @GuardedBy("this")
+    @Nullable
+    private volatile int[] hashTables;
+
+    MapHashTables(MapType mapType, Optional<int[]> hashTables)
+    {
+        this.mapType = mapType;
+        this.hashTables = hashTables.orElse(null);
+    }
+
+    public long getRetainedSizeInBytes()
+    {
+        return INSTANCE_SIZE + sizeOf(hashTables);
+    }
+
+    boolean isBuilt()
+    {
+        return hashTables != null;
+    }
 
     /**
-     * This method assumes that {@code keyBlock} has no duplicated entries (in the specified range)
+     * Returns the raw hash tables, which must not be modified.
+     * @throws IllegalStateException if the hash tables have not been built
      */
-    static void buildHashTable(Block keyBlock, int keyOffset, int keyCount, MapType mapType, int[] outputHashTable, int hashTableOffset, int hashTableSize)
+    int[] get()
+    {
+        if (hashTables == null) {
+            throw new IllegalStateException("hashTables are not built");
+        }
+        return hashTables;
+    }
+
+    /**
+     * Returns the raw hash tables, if they have been built.  The raw hash tables must not be modified.
+     * @throws IllegalStateException if the hash tables have not been built
+     */
+    Optional<int[]> tryGet()
+    {
+        return Optional.ofNullable(hashTables);
+    }
+
+    synchronized void growHashTables(int newSize)
     {
+        int[] hashTables = this.hashTables;
```

Comment:
This is line exists to improve the performance of accessing a volatile variable?

Reply:
Mostly it is there for my own sanity... it is easier to reason about.  Also, volatile access is a bit more expensive.